### ISSUE TITLE
A potential risk in curve which can be leveraged to make a cluster-level privilege escalation.

### ISSUE THREAD

**Issue by younaman** [opencurve/curve#2379]
Dear curve maintainers:
I am Nanzi Yang, and I find a potential risk in curve that can be
leveraged to make a cluster-level privilege escalation.

Detailed analysis:
1. The curve has one daemon set called csi-curve-plugin, which has one
pod on each node. The service account of this daemon set is
curve-csi-nodeplugin. It has a cluster role called
curve-csi-nodeplugin via cluster role binding. This cluster role has
"get list update" verbs of the "node" resource.
2. The curve has one deployment called csi-curve-plugin-provisioner,
which runs on worker nodes randomly. The service account of this
deployment is curve-csi-provisioner. It has a cluster role called
curve-external-provisioner-runner via cluster role binding. This
cluster role has "get list" verbs of the "secret" resources.

Attacking strategy:
If a malicious user can access a worker node:
1. He/she can leverage the csi-curve-plugin's service account to
update ALL other nodes with "node.kubernetes.io/unschedulable:
NoExecute" taint. Thus, the csi-curve-plugin-provisioner pod will be
forced to run on the malicious worker node.
2. After that, he/she can leverage the csi-curve-plugin-provisioner's
service account to list ALL secrets inside the Kubernetes cluster,
resulting in cluster-level privilege escalation.

Mitigation discussion:
1. The curve should remove the "update" verbs of the "node" resources.
2. The curve can use other mechanisms to retrain what node can be
updated by the daemon set. For example, the curve can use GateKeeper
to restrain the curve's daemon set can only patch its own worker
nodes. For another example, the curve can use the kubelet's config to
restrain the daemon set and can only "modify" the worker node which it
runs.
2. The curve can use the "resource name" to restrain what secrets can
be accessed by the deployments.
4. The curve can use the Role Binding, not the Cluster role binding,
to restrain that the deployment can only access secrets inside the
curve-csi-system Kubernetes namespace.

A few questions:
1. Is it a real issue in curve?
2. If it's a real issue, can curve fix this problem following my suggestions?
3. If it's a real issue, can curve give me a CVE number for awarding my efforts?

Looking forward to your reply.
Regards,
Nanzi Yang

**Comment by caoxianfei1** [comment]
Yes, but what you need to consider is that you have permission to access the cluster firstly, and generally Curve clusters or csi clusters are privatized deployments, and there may be no malicious attacks, of course, except for mistake by insiders outside.

**Comment by caoxianfei1** [comment]
@younaman Are you interested in the problem and want to try it ?

**Comment by younaman** [comment]
@caoxianfei1 I am very busy writing my paper right now. Can you or another maintainer fix this issue?
